File: ej4/src/main/java/com/mycompany/ej4/FilterChain.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package com.mycompany.ej4;

import java.util.ArrayList;

/**
 *
 * @author Usuario
 */
public class FilterChain {
    private final ArrayList<Filter> filters = new ArrayList<>();
    private Target target;
    
    public void addFilter(Filter filter){
        filters.add(filter);
    }
    
    public void setTarget(Target target){
        this.target = target;
    }
    
    public boolean execute(String data){
        for(Filter f : filters){
            //Si hay ya un filtro que falla la validación termina
            if(!f.execute(data)){
                System.out.println("Algún filtro ha fallado, intercepción terminada.");
                return false;
            }
        }
        
        if(target != null){
            target.execute(data);
        }
        
        return true;
    }
}

File: Practica_2/lib/Grupal/filterChain.dart
import 'package:practica_2/Grupal/filters.dart';
import 'package:practica_2/Grupal/target.dart';


class FilterChain {
  final List<Filter> filters = [];
  Target? target;

  void addFilter(Filter filter){
    filters.add(filter);
  }

  void setTarget(Target target){
    this.target = target;
  }

  int execute(String data){
    int success = 0 ;
    int resultado = 0 ;

    for(var f in filters){
      resultado = f.execute(data);
      if(resultado != 0){
        return resultado;
      }
    }

    if (target != null) {
      success = target!.execute(data);
    } else {
      throw Exception('Target no ha sido establecido antes de ejecutar la cadena de filtros.');
    }

    return success;
  }
}